src/Visitor.ts

```diff
@@ -0,0 +1,234 @@
+import * as ts from 'typescript'
+
+import { Counter } from './Counter'
+import { Descriptor } from './Descriptor'
+import { Input } from './Input'
+import * as lsif from './lsif'
+import { LsifSymbol } from './LsifSymbol'
+import { lsif_typed } from './main'
+import { Packages } from './Packages'
+import { Range } from './Range'
+
+export class Visitor {
+  private localCounter = new Counter()
+  private propertyCounters: Map<string, Counter> = new Map()
+  private localSymbolTable: Map<ts.Node, LsifSymbol> = new Map()
+  constructor(
+    public readonly checker: ts.TypeChecker,
+    public readonly input: Input,
+    public readonly document: lsif.lib.codeintel.lsif_typed.Document,
+    public readonly globalSymbolTable: Map<ts.Node, LsifSymbol>,
+    public readonly packages: Packages,
+    public readonly sourceFile: ts.SourceFile
+  ) {}
+  public index(): void {
+    this.visit(this.sourceFile)
+  }
+  private visit(node: ts.Node): void {
+    if (ts.isIdentifier(node)) {
+      const sym = this.checker.getSymbolAtLocation(node)
+      if (sym) {
+        this.visitIdentifier(node, sym)
+      }
+    }
+    ts.forEachChild(node, node => this.visit(node))
+  }
+
+  private visitIdentifier(node: ts.Identifier, sym: ts.Symbol): void {
```

Comment:
Nit: rename this parameter to `identifier`? It would make the code below more clearer, and also be less confusing since other methods use `node: ts.Node` whereas this has type `ts.Identifier`.